velox/exec/HashTable.cpp

```diff
@@ -614,21 +614,22 @@ void HashTable<ignoreNullKeys>::clear() {
 
 template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::checkSize(int32_t numNew) {
+  const int64_t numDistincts = numNew + numDistinct_;
   if (!table_ || !size_) {
     // Initial guess of cardinality is double the first input batch or at
     // least 2K entries.
     // numDistinct_ is non-0 when switching from HashMode::kArray to regular
     // hashing.
     auto newSize = std::max(
         (uint64_t)2048, bits::nextPowerOfTwo(numNew * 2 + numDistinct_));
-    if (numNew + numDistinct_ > rehashSize(newSize)) {
+    if (numDistincts > newSize || numDistincts > rehashSize(newSize)) {
```

Comment:
I don't think the first comparison has any effect since `rehashSize(x)` is `7/8 * x`